src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -52,31 +52,60 @@ public static TValue Deserialize<TValue>(string json, JsonSerializerOptions opti
         public static object Deserialize(string json, Type returnType, JsonSerializerOptions options = null)
         {
             if (json == null)
+            {
                 throw new ArgumentNullException(nameof(json));
+            }
 
             if (returnType == null)
+            {
                 throw new ArgumentNullException(nameof(returnType));
+            }
 
-            return ParseCore(json, returnType, options);
-        }
-
-        private static object ParseCore(string json, Type returnType, JsonSerializerOptions options = null)
-        {
             if (options == null)
             {
                 options = JsonSerializerOptions.s_defaultOptions;
             }
 
-            // todo: use an array pool here for smaller requests to avoid the alloc?
-            byte[] jsonBytes = JsonReaderHelper.s_utf8Encoding.GetBytes(json);
-            var readerState = new JsonReaderState(options.GetReaderOptions());
-            var reader = new Utf8JsonReader(jsonBytes, isFinalBlock: true, readerState);
-            object result = ReadCore(returnType, options, ref reader);
+            object result;
+            byte[] tempArray = null;
 
-            if (reader.BytesConsumed != jsonBytes.Length)
+            int maxByteCount;
+            if (json.Length > JsonConstants.MaxEscapedTokenSize / JsonConstants.MaxExpansionFactorWhileTranscoding)
+            {
+                // Get the actual byte count in order to handle large input.
+                maxByteCount = JsonReaderHelper.GetUtf8ByteCount(json);
```

Comment:
Implicit cast from string to `ROS<char>` isn't available on netfx. Use AsSpan.
```suggestion
                maxByteCount = JsonReaderHelper.GetUtf8ByteCount(json.AsSpan());
```